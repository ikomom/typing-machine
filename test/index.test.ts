import type { Diff } from 'diff-match-patch-typescript'
import { DiffMatchPatch, DiffOperation } from 'diff-match-patch-typescript'
import { describe, expect, it } from 'vitest'

const input = `
import { describe, expect, it } from 'vitest'
import {ll} from './one'

describe('should', () => {
  it('exported', () => {
    expect(33)
    console.log('hello')
  })
})
`

const output = `
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33).toEqual(1)
  })
})
`

interface InsertPatch {
  type: 'insert'
  // 删除开始位置
  from: number
  text: string
}
interface RemovalPatch {
  type: 'removal'
  // 删除开始位置
  from: number
  length: number
}

type Patch = InsertPatch | RemovalPatch

function calculatePatches(diffs: Diff[]): Patch[] {
  const patches: Patch[] = []
  // 当前文字的index
  let cursor = 0
  for (const change of diffs) {
    const [operation, text] = change
    const offset = text.length
    if (operation === DiffOperation.DIFF_EQUAL) {
      cursor += offset
    }
    else if (operation === DiffOperation.DIFF_DELETE) {
      patches.push({
        type: 'removal',
        from: cursor,
        length: offset,
        // // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // // @ts-expect-error
        // text, // TODO: 测试用, 待移除
      })
      // 删除时，游标还是一样，所以不用移动，一下代码不用
      // cursor -= offset
    }
    else if (operation === DiffOperation.DIFF_INSERT) {
      patches.push({
        type: 'insert',
        from: cursor,
        text,
      })
      // 添加时, 游标 = 之前游标 + 添加长度
      cursor += offset
    }
    else {
      throw new Error('unknown change type')
    }
  }
  return patches
}

function diffString(a: string, b: string): Diff[] {
  const diff = new DiffMatchPatch()
  const res = diff.diff_main(a, b)
  diff.diff_cleanupSemantic(res)
  return res
}

function applyPatches(input: string, patches: Patch[]) {
  let output = input
  // 每次只关注添加、删除的位置, 并且拼接起来
  for (const patch of patches) {
    const head = output.slice(0, patch.from)
    if (patch.type === 'insert') {
      const tail = output.slice(patch.from)
      output = head + patch.text + tail
    }
    else if (patch.type === 'removal') {
      const tail = output.slice(patch.from + patch.length)
      output = head + tail
    }
  }
  return output
}

describe('should', () => {
  it('diff', () => {
    const delta = diffString(input, output)
    expect(delta).toMatchSnapshot('delta')
    const patches = calculatePatches(delta)
    expect(patches).toMatchSnapshot('patches')
    const applied = applyPatches(input, patches)
    expect(applied).toMatchSnapshot('output')
    expect(applied).toEqual(output)
  })
})
