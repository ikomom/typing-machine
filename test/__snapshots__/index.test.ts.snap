// Vitest Snapshot v1

exports[`index > diff > delta 1`] = `
[
  [
    0,
    "
import { describe, expect, it } from 'vitest'",
  ],
  [
    -1,
    "
import {ll} from './one'",
  ],
  [
    0,
    "

describe('should', () => {
  it('exported', () => {
    expect(33)",
  ],
  [
    -1,
    "
    console.log('hello'",
  ],
  [
    1,
    ".toEqual(1",
  ],
  [
    0,
    ")
  })
})
",
  ],
]
`;

exports[`index > diff > output 1`] = `
"
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33).toEqual(1)
  })
})
"
`;

exports[`index > diff > patches 1`] = `
[
  {
    "from": 46,
    "length": 25,
    "type": "removal",
  },
  {
    "from": 114,
    "length": 24,
    "type": "removal",
  },
  {
    "from": 114,
    "text": ".toEqual(1",
    "type": "insert",
  },
]
`;
